vetomint2/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+mod progress;
+mod state;
+
+use serde::{Deserialize, Serialize};
+
+/// An index of the validator, which is for a single height. (Mapping from the actual public key to the index may differ for different heights.)
+pub type ValidatorIndex = usize;
+/// An identifier of the block, which is uniquely mapped to a block. Like `ValidatorIndex`, it is for a single height. (Mapping from the actual block to the index may differ for different heights.)
+pub type BlockIdentifier = usize;
+/// A round.
+pub type Round = usize;
+/// A voting power.
+pub type VotingPower = u64;
+/// A UNIX timestamp measured in milliseconds.
+pub type Timestamp = i64;
+
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
+pub struct ConsensusParams {
+    pub timeout_ms: u64,
+    pub repeat_round_for_first_leader: usize,
+}
+
+/// An event that (potentially) triggers a state transition of `StateMachine`.
+///
+/// Note that there is no cryptography-related info here, because it's
+/// the lower layer's responsibility to verifiy and refine the raw messages (containing such cryptography-related info) into this abstracted data.
+/// Also all the identifiers (for blocks and validators) become integer indices here, and
+/// the lower layer will keep the mapping from the actual data to the indices.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ConsensusEvent {
+    /// Signals to start the process
+    Start,
+    /// Informs that the node has received a block proposal.
+    BlockProposalReceived {
+        proposal: BlockIdentifier,
+        /// Whether this proposal is valid
+        valid: bool,
```

Comment:
Is "valid" determined at the lower layer after each node receives the block?

Reply:
yes.